controller/model/asynchronous_eventer.go

```diff
@@ -0,0 +1,280 @@
+	BucketName  string
+	Interval    time.Duration
+	closeNotify <-chan struct{}
+	stopNotify  chan struct{}
+
+	env Env
+
+	running    concurrenz.AtomicBoolean
+	processing concurrenz.AtomicBoolean
+	batchSize  int
+}
+
+var _ AsynchronousEventer = &AsynchronousEventerBbolt{}
+
+// NewAsynchronousEventerBbolt creates a new bbolt backed asynchronous eventer that will check for new events at the given interval
+// or when triggered. On each interval/trigger, the number of events processed is determined by batchSize.
+func NewAsynchronousEventerBbolt(env Env, interval time.Duration, batchSize int) *AsynchronousEventerBbolt {
+	return &AsynchronousEventerBbolt{
+		Interval:  interval,
+		env:       env,
+		batchSize: batchSize,
+	}
+}
+
+func (a *AsynchronousEventerBbolt) AddEvent(eventType string, data []byte) {
+	err := a.env.GetDbProvider().GetDb().Update(func(tx *bbolt.Tx) error {
+		ctx := boltz.NewMutateContext(tx)
+		event := &persistence.AsynchronousEvent{
+			BaseExtEntity: boltz.BaseExtEntity{
+				Id:        cuid.New(),
+				CreatedAt: time.Now(),
+				UpdatedAt: time.Now(),
+			},
+			Type: eventType,
+			Data: data,
+		}
+
+		return a.env.GetStores().AsynchronousEvent.Create(ctx, event)
+	})
+
+	if err != nil {
+		pfxlog.Logger().WithError(err).Error("error adding event for AsynchronousEventerBbolt")
+	}
+}
+
+func (a *AsynchronousEventerBbolt) AddListener(eventType string, listener EventListenerFunc) {
+	a.handlerMap.Upsert(eventType, nil, func(exist bool, valueInMap interface{}, newValue interface{}) interface{} {
+		var handlers []EventListenerFunc
+		if exist {
+			handlers = valueInMap.([]EventListenerFunc)
+		}
+
+		handlers = append(handlers, listener)
+		return handlers
+	})
+}
+
+func (a *AsynchronousEventerBbolt) Start(closeNotify <-chan struct{}) error {
+	if !a.running.CompareAndSwap(false, true) {
+		return errors.New("already started")
+	}
+	a.stopNotify = make(chan struct{}, 0)
+	a.closeNotify = closeNotify
+
+	go a.run()
+
+	return nil
+}
+
+func (a *AsynchronousEventerBbolt) run() {
+
+	for {
+		select {
+		case <-a.closeNotify:
+			a.running.Set(false)
+			return
+		case <-a.stopNotify:
+			a.running.Set(false)
+			return
+		case <-time.After(a.Interval):
+			a.process()
+		}
+	}
+}
+
+func (a *AsynchronousEventerBbolt) Stop() error {
+	if a.running.CompareAndSwap(true, false) {
+		close(a.stopNotify)
+		return nil
+	}
+
+	return errors.New("not started")
+}
+
+func (a *AsynchronousEventerBbolt) Trigger() error {
+	if a.running.Get() {
+		go a.process()
+		return nil
+	}
+
+	return errors.New("not started")
+}
+
+func (a *AsynchronousEventerBbolt) deleteEvent(id string) error {
+	return a.env.GetDbProvider().GetDb().Update(func(tx *bbolt.Tx) error {
+		ctx := boltz.NewMutateContext(tx)
+		return a.env.GetStores().AsynchronousEvent.DeleteById(ctx, id)
+	})
+}
+
+func (a *AsynchronousEventerBbolt) process() {
+	if !a.processing.CompareAndSwap(false, true) {
+		return //already running
+	}
+
```

Comment:
Do we need this defer if we have the defer where the handlers are called below?

Reply:
We don't need the panic resolution but I do want to make sure processing is set to false.

Removed.